solutions/USACO open18-familytree.mdx

```diff
@@ -0,0 +1,127 @@
+---
+id: usaco-833
+source: USACO Bronze 2018 Open
+title: Family Tree
+author: Danh Ta Chi Thanh
+---
+
+## Solution
+
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_family_bronze_open18.html)
+
+## Alternative Solution - BFS
+
+This problem can also be solved by using BFS to find the common ancestors and the distance between the two cows and their nearest common ancestor.
+
+<Warning>
+
+No Bronze / Silver problem has required knowledge of BFS, so feel free to skip this solution if you find it unfamiliar and come back later.
+
+</Warning>
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h> // see /general/running-code-locally
+using namespace std;
+
+//vectors
+using vi = vector<int>;
+#define all(x) begin(x), end(x)
+#define sz(x) int((x).size())
+#define pb push_back
+
+//pairs
+using pi = pair<int,int>;
+#define f first
+#define s second
+#define mp make_pair
+
+void setIO(string name = "") {
+	cin.tie(0)->sync_with_stdio(0); // see /general/fast-io
+	if (sz(name)) {
+		freopen((name+".in").c_str(), "r", stdin); // see /general/io
+		freopen((name+".out").c_str(), "w", stdout);
+	}
+}
+
+int main() {
+	setIO("family");
+	int N;
+	string c[3];
+	cin >> N >> c[1] >> c[2];
+	vector<pair<string,int>> dist[3];
+	dist[1].pb(mp(c[1],0));
+	dist[2].pb(mp(c[2],0));
+
+	vector<pair<string,string>> edge(N); //edge list
+
+	for (int i = 0; i < N; ++i) {
+		string u, v;
+		cin >> v >> u;
+		// If v is a mother of u, then there is an edge from node u to node v.
+		edge[i] = mp(u,v);
+	}
+
+	for (int k = 1; k <= 2; ++k) {
+		queue<string> q;
+		q.push(c[k]);
+		// BFS on directed tree
+		// to find the ancestors of both cows and the distance to them
+		while (!q.empty()) {
+			string node = q.front();
+			q.pop();
+			for (auto& E : edge) if (E.f == node) {

```

Comment:
This is potentially bad code practice... You should use something that points E.f to E.s instead of looping through all edges. Pretend the constraints are actually high

Reply:
maybe 'map<string,string>`?

Reply:
Yes, I suppose that's fine.

Reply:
Wouldn't it be simpler to match each string to an integer and just stick to arrays?